### ISSUE TITLE
Arithmetic expression creates SC2125 false positive

### ISSUE THREAD

**Issue by ronburk** [koalaman/shellcheck#914]
See related difficulties with the integer attribute causing false positives in #189
#### For bugs
- Rule Id SC2125 (also SC2034)
- My shellcheck version 0.4.6
- [X ] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:
```
#!/bin/bash
function foo(){
    local -i iArgc="$1"
    : ...
    local -i ArgCount=BASH_ARGC[iArgc]
    printf "%d\n" "$ArgCount"
}
foo 0

```
#### Here's what shellcheck currently says:
```
In test3.sh line 3:
    local -i iArgc="$1"
             ^-- SC2034: iArgc appears unused. Verify it or export it.


In test3.sh line 5:
    local -i ArgCount=BASH_ARGC[iArgc]
                      ^-- SC2125: Brace expansions and globs are literal in assignments. Quote it or use an array.

```
#### Here's what I wanted or expected to see:
As far as I can see, both are false positives. The SC2034 problem is perhaps the same as referenced in #189. Guessing without reading the code, perhaps the SC2125 arises from simply not treating variables with the -i attribute any differently, in which case the brackets are mistaken for glob pattern characters. Ironically, that makes the "or use an array" message extra confusing, since of course I **am** using an array. :-)

**Comment by kwshi** [comment]
Having this issue as well.  As a side note, when is `[]` square brackets used for glob patterns, anyway?  Is there even a reason for SC2125 to flag square brackets, independent of the arithmetic expression issue?

**Comment by kurahaupo** [comment]
@kwshi 

> As a side note, when is `[]` square brackets used for glob patterns, anyway?

The short answer is "always, except when there's an defined exception".

Right up front it's worth noting that if a glob expansion does not match anything, it will be left unchanged, so it might _seem_ like glob expansions only happen occasionally.

The exceptions include:

* On the right hand side of a bare assignment
* On the right hand side of an assignment that's an argument to an unquoted _special_ built-in command
  * `declare`
  * `export`
  * `local`
  * `readonly`
* In the selector word between `case` and `in`
* Inside `((`...`))`
* Inside `[[`...`]]` other than on the right side of `=` or `==` or `=~`

(and not quoted, of course).

Compare these two uses of the `declare` built-in command:
```
$ touch a=b0
$ declare a=b[0] ; echo "[$a]"    # no globbing because declare is special
[b[0]]
$ 'declare' a=b[0] ; echo "[$a]"  # globbing because 'declare' is not special
[b0]
```

**Comment by LLyaudet** [comment]
Hello,

I have the same problem:
```
#!/usr/bin/env bash

foo(){
  declare -gi bar
  bar=7+$1*4
  export bar
}
```

Bash manual says that the parentheses are not needed after using declare -i.
But shellcheck treats bar as if it was a string variable.

**Comment by kurahaupo** [comment]
@LLyaudet just to be clear, the `-g` option on `declare` affects _only_ that statement. It does not affect uses of `bar` elsewhere in the function, unless it happens to refer to the top-level global variable. Be aware that `local` creates a scope that extends into any subsidiary functions; it's _not_ limited to just the current function. Also be aware that `-i` only affects assignment, not expansion. If a variable already has a non-numeric value, setting `-i` on it won't change its value.

This combination is demonstrated when you run:
```
#!/dont/use/env/./change/this/line/to/the/path/to/your/bash
 
# shellcheck enable=all
 
foo() {
  declare -gi bar
  bar=7+$1*4
  export bar
  echo "$FUNCNAME:$LINENO:bar=$bar"
  echo "$FUNCNAME:$LINENO:((bar))=$((bar))"
}
 
zot() {
  local bar
  foo 40+2
  echo "$FUNCNAME:$LINENO:bar=$bar"
  echo "$FUNCNAME:$LINENO:((bar))=$((bar))"
}
 
bar=Hello
echo "main:$LINENO:bar=$bar"
zot
echo "main:$LINENO:bar=$bar"
bar=9*3
echo "main:$LINENO:bar=$bar"
```
and get output that looks like:
```
main:21:bar=Hello
foo:9:bar=7+40+2*4
foo:10:((bar))=55
zot:16:bar=7+40+2*4
zot:17:((bar))=55
main:23:bar=Hello
main:25:bar=27
```

**Comment by LLyaudet** [comment]
Hello @kurahaupo :),

Thank you very much for these remarks.
I did know that "Be aware that local creates a scope that extends into any subsidiary functions;",
and that's why I used LFBFL_prefix for internal variables in my code to avoid unexpected side-effects if someone else uses some functions in my code.
But I didn't realize that it could shadow a `declare -i or -g` below.
I tested your code with and without the `local bar` line, and 42 instead of 40 + 2.
`declare -g` without local does reach the main scope and all outer scopes which is my goal when using it.
I will probably remove the disable for SC2125 in my code and use double parentheses even if it works without it.

I think the variables scopes handling is probably the worst "feature" of shell.
The logic behind it is obscure to me.
If you know why it works this way and can provide references, I would be happy to learn.

Regarding using env or just /bin/bash,
I followed the main answer here:
https://stackoverflow.com/questions/21612980/why-is-usr-bin-env-bash-superior-to-bin-bash/78413334
Even if I had upvoted both the main answer and the answer saying that it is not useful for bash.
Let me know if you have comments regarding the answers on SO.

Best regards,
    Laurent Lyaudet

**Comment by LLyaudet** [comment]
@kurahaupo 
Do you think that the Bash manual should be improved ?
My first opinion before adding my first comment in this issue was driven by the sentence:
"If the variable has its integer attribute set, then value is evaluated as an arithmetic expression even if the $((…)) expansion is not used (see Arithmetic Expansion)."
Maybe this sentence should be followed by a warning saying that the integer attribute may not be set, even after using -i.
And some example similar to yours be added in the Bash manual.

**Comment by kurahaupo** [comment]
@LLyaudet yes I think there's much that could be improved in the Bash manual.

It needs to be made clearer that the numeric evaluation occurs on assignment and not on expansion.

The `-g` option needs a really big warning that it does _not_ have any effect beyond the `local` or `declare` statement.

Sadly the Bash maintainer is very unlikely to add counter-examples to the manual.

**Comment by LLyaudet** [comment]
I found another very annoying "bug/feature" that I can explain with your code example:<br>
 the `declare -ig bar` is used when bar is not defined at all or defined before in main, but it is not used at all when there is a `local bar` or `declare bar` definition "above" in `zot`,<br>
hence if you write a single line like `declare -ig bar=7+$1*4` the assignment is droped...<br>
So you're advised to never do that which can be a real problem when you have variable variable names (not a typo on double variable).
